language_service/src/state.rs

```diff
@@ -0,0 +1,465 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#[cfg(test)]
+mod tests;
+
+use super::compilation::Compilation;
+use super::protocol::{DiagnosticUpdate, NotebookMetadata};
+use crate::protocol::WorkspaceConfigurationUpdate;
+use log::{error, trace};
+use miette::Diagnostic;
+use qsc::compile::Error;
+use qsc::target::Profile;
+use qsc::PackageType;
+use qsc_project::{FileSystemAsync, JSFileEntry};
+use rustc_hash::{FxHashMap, FxHashSet};
+use std::{cell::RefCell, fmt::Debug, future::Future, mem::take, pin::Pin, rc::Rc, sync::Arc};
+
+/// the desugared return type of an "async fn"
+type PinnedFuture<T> = Pin<Box<dyn Future<Output = T>>>;
+
+/// represents a unary async fn where `Arg` is the input
+/// parameter and `Return` is the return type. The lifetime
+/// `'a` represents the lifetime of the contained `dyn Fn`.
+type AsyncFunction<'a, Arg, Return> = Box<dyn Fn(Arg) -> PinnedFuture<Return> + 'a>;
+
+#[derive(Default)]
+pub(super) struct CompilationState {
+    /// A `CompilationUri` is an identifier for a unique compilation.
+    /// It is NOT required to be a uri that represents an actual document.
+    ///
+    /// For single Q# documents, the `CompilationUri` is the same as the
+    /// document uri.
+    ///
+    /// For notebooks, the `CompilationUri` is the notebook uri.
+    ///
+    /// The `CompilatinUri` is used when compilation-level errors get reported
+    /// to the client. Compilation-level errors are defined as errors without
+    /// an associated source document.
+    ///
+    /// `PartialConfiguration` contains configuration overrides for this
+    /// compilation, explicitly specified through a project manifest (not currently implemented)
+    /// or notebook metadata.
+    compilations: FxHashMap<CompilationUri, (Compilation, PartialConfiguration)>,
+    /// All the documents that we were told about by the client.
+    ///
+    /// This map doesn't necessarily contain ALL the documents that
+    /// make up a compilation - only the ones that are currently open.
+    open_documents: FxHashMap<DocumentUri, OpenDocument>,
+}
+
+type CompilationUri = Arc<str>;
+type DocumentUri = Arc<str>;
+
+#[derive(Debug)]
+struct OpenDocument {
+    /// This version is the document version provided by the client.
+    /// It increases strictly with each text change, though this knowledge should
+    /// not be important. The version is only ever used when publishing
+    /// diagnostics to help the client associate the list of diagnostics
+    /// with a snapshot of the document.
+    pub version: u32,
+    pub compilation: CompilationUri,
+}
+
+#[derive(Debug, Clone)]
+struct Configuration {
```

Comment:
Impl `Copy` instead of `Clone`?

Reply:
Why? `Copy` requires `Clone` so I'd have to implement both. Nobody requires `Copy` on this struct at the moment.

Reply:
Finally got it. The phrase "instead of" threw me off - you need to derive `Copy` _and_ `Clone` . Anyway it's a good change thank you!